### ISSUE TITLE
tfd.Categorical.log_prob return dtype given list of logits

### ISSUE THREAD

**Issue by chrism0dwk** [tensorflow/probability#973]
In `tfd.Categorical`, the `log_prob` method derives its return type from the `logits` (or `probs`) constructor argument, as in the following: 
```
logits32 = tf.constant([1,2,3,4], dtype=tf.float32)
tfd.Categorical(logits=logits).log_prob(2).dtype == tf.float32 # True

logits64 = tf.constant([1,2,3,4], dtype=tf.float64)
tfd.Categorical(logits=logits).log_prob(2).dtype == tf.float64 # True
```
However, if a list (batch) of `[logits]` is supplied, an unexpected type demotion occurs such that the return value of `log_prob` defaults `tf.float32`:
```
logits64 = tf.constant([1,2,3,4], dtype=tf.float64)
tfd.Categorical(logits=[logits]).log_prob(2).dtype == tf.float64 # False
```

**Comment by foabodo** [comment]
This is a hack, but you can use this equivalent to unblock yourself:
`tfd.Multinomial(logits=[logits64], total_count=1).log_prob([2]).dtype == tf.float64`
